File: sdk/javascript/src/explain.js
/**
 * Generate a human-readable narrative from a FoodBlock graph.
 * Walks the provenance tree and renders a plain-English explanation.
 *
 * Usage:
 *   const story = await fb.explain(breadHash, resolve)
 *   // "Sourdough ($4.50) by Green Acres Bakery..."
 */

const TYPE_VERBS = {
  'transform': 'made by',
  'transfer': 'via',
  'observe': '',
}

const OBSERVE_LABELS = {
  'observe.review': 'review',
  'observe.certification': 'certified by',
  'observe.inspection': 'inspected by',
  'observe.reading': 'reading from',
  'observe.scan': 'scanned by',
}

/**
 * Generate a narrative for a block and its provenance.
 *
 * @param {string} hash - Hash of the block to explain
 * @param {function} resolve - async (hash) => block | null
 * @param {object} [opts] - { maxDepth: number (default 10) }
 * @returns {string} Human-readable narrative
 */
async function explain(hash, resolve, opts = {}) {
  const maxDepth = opts.maxDepth || 10
  const block = await resolve(hash)
  if (!block) return `Block not found: ${hash}`

  const parts = []
  const visited = new Set()

  await buildNarrative(block, resolve, parts, visited, 0, maxDepth)

  return parts.join(' ')
}

async function buildNarrative(block, resolve, parts, visited, depth, maxDepth) {
  if (!block || visited.has(block.hash) || depth > maxDepth) return
  visited.add(block.hash)

  const name = block.state.name || block.state.title || block.type
  const baseType = block.type.split('.')[0]

  // Describe the block itself
  if (depth === 0) {
    let desc = name
    if (block.state.price) desc += ` ($${block.state.price})`
    if (block.state.rating) desc += ` (${block.state.rating}/5)`
    parts.push(desc + '.')
  }

  // Follow key refs to build the story
  const refs = block.refs || {}

  // Actor refs (seller, buyer, author, operator)
  for (const role of ['seller', 'buyer', 'author', 'operator', 'producer']) {
    if (refs[role]) {
      const actor = await resolve(refs[role])
      if (actor && actor.state.name && !visited.has(actor.hash)) {
        visited.add(actor.hash)
        if (depth === 0) {
          parts.push(`By ${actor.state.name}.`)
        }
      }
    }
  }

  // Input/source refs (provenance)
  for (const role of ['inputs', 'source', 'origin', 'input']) {
    if (refs[role]) {
      const refHashes = Array.isArray(refs[role]) ? refs[role] : [refs[role]]
      const names = []
      for (const h of refHashes) {
        const dep = await resolve(h)
        if (dep && dep.state.name) {
          let depDesc = dep.state.name
          // Check for source actor
          const depSource = dep.refs && (dep.refs.seller || dep.refs.source || dep.refs.producer)
          if (depSource) {
            const sourceActor = await resolve(depSource)
            if (sourceActor && sourceActor.state.name) {
              depDesc += ` (${sourceActor.state.name})`
            }
          }
          names.push(depDesc)
        }
      }
      if (names.length > 0) {
        parts.push(`Made from ${names.join(', ')}.`)
      }
    }
  }

  // Certifications
  if (refs.certifications) {
    const certHashes = Array.isArray(refs.certifications) ? refs.certifications : [refs.certifications]
    for (const h of certHashes) {
      const cert = await resolve(h)
      if (cert && cert.state.name) {
        let certDesc = `Certified: ${cert.state.name}`
        if (cert.state.valid_until) certDesc += ` (expires ${cert.state.valid_until})`
        parts.push(certDesc + '.')
      }
    }
  }

  // Update chain
  if (refs.updates && !visited.has(refs.updates)) {
    const prev = await resolve(refs.updates)
    if (prev) {
      const prevName = prev.state.name || prev.type
      if (prev.state.price && block.state.price && prev.state.price !== block.state.price) {
        parts.push(`Updated from $${prev.state.price}.`)
      }
    }
  }

  // Tombstone
  if (block.state.tombstoned) {
    parts.push('This block has been erased.')
  }
}

module.exports = { explain }

File: sdk/go/explain.go
package foodblock

import "fmt"

// Explain generates a human-readable narrative for a block and its provenance.
func Explain(hash string, resolve func(string) *Block, maxDepth int) string {
	if maxDepth <= 0 {
		maxDepth = 10
	}
	block := resolve(hash)
	if block == nil {
		return fmt.Sprintf("Block not found: %s", hash)
	}

	visited := make(map[string]bool)
	parts := buildNarrative(block, resolve, visited, 0, maxDepth)
	result := ""
	for i, p := range parts {
		if i > 0 {
			result += " "
		}
		result += p
	}
	return result
}

func buildNarrative(block *Block, resolve func(string) *Block, visited map[string]bool, depth, maxDepth int) []string {
	if block == nil || visited[block.Hash] || depth > maxDepth {
		return nil
	}
	visited[block.Hash] = true

	name := ""
	if n, ok := block.State["name"].(string); ok {
		name = n
	} else if t, ok := block.State["title"].(string); ok {
		name = t
	} else {
		name = block.Type
	}

	var parts []string

	if depth == 0 {
		desc := name
		if price, ok := block.State["price"].(float64); ok {
			desc += fmt.Sprintf(" ($%.2f)", price)
		}
		if rating, ok := block.State["rating"].(float64); ok {
			desc += fmt.Sprintf(" (%.0f/5)", rating)
		}
		parts = append(parts, desc+".")
	}

	refs := block.Refs

	// Actor refs
	for _, role := range []string{"seller", "buyer", "author", "operator", "producer"} {
		if refHash, ok := refs[role].(string); ok {
			actor := resolve(refHash)
			if actor != nil && !visited[actor.Hash] {
				if actorName, ok := actor.State["name"].(string); ok {
					visited[actor.Hash] = true
					if depth == 0 {
						parts = append(parts, "By "+actorName+".")
					}
				}
			}
		}
	}

	// Input/source refs
	for _, role := range []string{"inputs", "source", "origin", "input"} {
		ref, ok := refs[role]
		if !ok {
			continue
		}
		var refHashes []string
		switch v := ref.(type) {
		case string:
			refHashes = []string{v}
		case []interface{}:
			for _, item := range v {
				if s, ok := item.(string); ok {
					refHashes = append(refHashes, s)
				}
			}
		}

		var names []string
		for _, h := range refHashes {
			dep := resolve(h)
			if dep == nil {
				continue
			}
			if depName, ok := dep.State["name"].(string); ok {
				depDesc := depName
				for _, srcRole := range []string{"seller", "source", "producer"} {
					if srcHash, ok := dep.Refs[srcRole].(string); ok {
						srcActor := resolve(srcHash)
						if srcActor != nil {
							if srcName, ok := srcActor.State["name"].(string); ok {
								depDesc += " (" + srcName + ")"
							}
						}
						break
					}
				}
				names = append(names, depDesc)
			}
		}
		if len(names) > 0 {
			joined := ""
			for i, n := range names {
				if i > 0 {
					joined += ", "
				}
				joined += n
			}
			parts = append(parts, "Made from "+joined+".")
		}
	}

	// Certifications
	if certRef, ok := refs["certifications"]; ok {
		var certHashes []string
		switch v := certRef.(type) {
		case string:
			certHashes = []string{v}
		case []interface{}:
			for _, item := range v {
				if s, ok := item.(string); ok {
					certHashes = append(certHashes, s)
				}
			}
		}
		for _, h := range certHashes {
			cert := resolve(h)
			if cert == nil {
				continue
			}
			if certName, ok := cert.State["name"].(string); ok {
				certDesc := "Certified: " + certName
				if validUntil, ok := cert.State["valid_until"].(string); ok {
					certDesc += " (expires " + validUntil + ")"
				}
				parts = append(parts, certDesc+".")
			}
		}
	}

	// Tombstone
	if tombstoned, ok := block.State["tombstoned"].(bool); ok && tombstoned {
		parts = append(parts, "This block has been erased.")
	}

	return parts
}
